File: Algorithm/Code_Py/4.最长公共子序列/最长公共子序列.py
#实现最长公共子序列

#导入相应的库
import numpy as np

#定义状态数组
f = np.zeros((101,101))

def DP(t1,t2):
    for i in range(1,len(t1)+1):
        for j in range(1,len(t2)+1):
            if t1[i-1] == t2[j-1]:
                f[i][j] = f[i-1][j-1] + 1
            else:
                f[i][j] = max(f[i-1][j],f[i][j-1])
    return f[len(t1)][len(t2)]



if __name__ == "__main__":
    text1 = "ADABEC"
    text2 = "DBDCA"
    answer = DP(text1,text2)
    print("最长公共子序列的长度为：",answer)

File: Algorithm/Code/5.最长公共子序列/最长公共子序列.cpp
//最长公共子序列
#include<iostream>
#include<string>
using namespace std; 
int f[101][101]; //状态数组



int DP(string t1,string t2){
	for(int i = 1;i<=t1.size();i++){
		for(int j = 1;j <= t2.size();j++){
			if(t1[i - 1] == t2[j - 1]) f[i][j] = f[i - 1][j - 1] + 1;
			else
				f[i][j] = max(f[i][j - 1],f[i - 1][j]);
		}
	}
	return f[t1.size()][t2.size()];
}



int main(){
	string text1 = "ADABEC";
	string text2 = "DBDCA";
	int ans = DP(text1,text2);
	printf("最长公共子序列为：%d\n",ans);
	return 0;
} 
